### PR TITLE
🐛 Source Github: log the error message for streams that are restricted for OAuth

### PR DESCRIPTION
**Pull request by bazarnov**
## What
Resolving On-Call: https://github.com/airbytehq/oncall/issues/117

## How
- added the Full Error message with the description about what caused the `403 Forbidden` for certain streams, like `Teams`
- added more scopes for OAuth Flow implementation for `source-github` (GitHubOAuthFlow.java)
- added `config_oauth` into SAT tests + created new secret for that, this part wasn't tested before

## 🚨 User Impact 🚨
Users must update the connector version up to `0.2.16`, in some cases users should proceed the `re-authentication` for their GitHub connector, check the  `Note` bellow.

### Note:

`Re-authenticate` the connector is required if:
- for all users who has the issues with `403 Forbidden` for certain streams, on versions < `0.2.16` using Airbyte-Cloud.

`Otherwise`:
- no action required

## Pre-merge Checklist
Expand the relevant checklist and delete the others. 

<details><summary> <strong> Updating a connector </strong></summary>
<p>
#### Community member or Airbyter

- [X] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [X] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [ ] Connector's `README.md`
    - [X] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [X] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the new connector version is published, connector version bumped in the seed directory as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</p>
</details>

### CODE REVIEW

**Comment by ChristopheDuong** [comment]
All configurations with previously configured github source would have to re-auth for new scopes?

**Comment by bazarnov** [comment]
> All configurations with previously configured github source would have to re-auth for new scopes?

**Not necessary:**
- if sync was successful previously, or if the PAT is used as authentication option. 

**Necessary**
- for all users who has the issues with `403 Forbidden` for certain streams, on versions < `0.2.16` using Airbyte-Cloud.

File: airbyte-oauth/src/main/java/io/airbyte/oauth/flows/GithubOAuthFlow.java [hunk 1]
```diff
@@ -47,13 +60,13 @@ protected String formatConsentUrl(final UUID definitionId,
...
+
     try {
       return new URIBuilder(AUTHORIZE_URL)
           .addParameter("client_id", clientId)
           .addParameter("redirect_uri", redirectUrl)
-          .addParameter("scope", SCOPES)
-          .addParameter("state", getState())
-          .build().toString();
+          // we need to keep scopes in the format of: < scope1%20scope2:sub_scope%20scope3 >
+          .build().toString() + "&scope=" + getScopes() + "&state=" + getState();
```
Comment (grubberr) [inline airbyte-oauth/src/main/java/io/airbyte/oauth/flows/GithubOAuthFlow.java:69]: Why we extracted `getState` from URIBuilder ?

File: airbyte-integrations/connectors/source-github/source_github/streams.py [hunk 2]
```diff
@@ -81,23 +81,33 @@ def backoff_time(self, response: requests.Response) -> Union[int, float]:
...
+        # get out the stream_slice parts for later use.
+        organisation = stream_slice.get("organization", "")
+        repository = stream_slice.get("repository", "")
+        # Reading records while handling the errors
         try:
             yield from super().read_records(stream_slice=stream_slice, **kwargs)
         except HTTPError as e:
-            error_msg = str(e)
-
+            error_msg = str(e.response.json().get("message"))
```
Comment (grubberr) [inline airbyte-integrations/connectors/source-github/source_github/streams.py:91]: Are any probability that error will be not in json format?

**Review by grubberr — APPROVED** [review]

File: airbyte-oauth/src/main/java/io/airbyte/oauth/flows/GithubOAuthFlow.java [hunk 3]
```diff
@@ -30,7 +32,18 @@
...
+      "repo",
+      "read:org",
+      "read:repo_hook",
+      "read:user",
+      "read:discussion",
+      "workflow");
+
+  public String getScopes() {
+    return String.join("%20", SCOPES);
+  }
```
Comment (antixar) [inline airbyte-oauth/src/main/java/io/airbyte/oauth/flows/GithubOAuthFlow.java:46]: "%20" is the code " " (space)
Try to concatenate this array by simple space and use the standard URIBuilder for generation of links

**Review by antixar — COMMENTED** [review]
Please try to check my assumption

File: airbyte-oauth/src/main/java/io/airbyte/oauth/flows/GithubOAuthFlow.java (hunk 3 above)
Comment (bazarnov) [reply airbyte-oauth/src/main/java/io/airbyte/oauth/flows/GithubOAuthFlow.java:46]: By using default UriBuilder flow, the others symbols like ` : ` are also encoded. To grant the permissions correctly, we should keep this method of encoding, like implemented. 
Also, refer to this article about oauth scopes:  https://docs.github.com/en/developers/apps/building-oauth-apps/scopes-for-oauth-apps (Note section, after scopes table)

File: airbyte-integrations/connectors/source-github/source_github/streams.py (hunk 2 above)
Comment (bazarnov) [reply airbyte-integrations/connectors/source-github/source_github/streams.py:91]: No, the response should be always in `json` format.

**Pull request merged**